tower-http/src/services/fs/serve_dir.rs

```diff
@@ -315,49 +346,112 @@ impl Future for ResponseFuture {
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         match &mut self.inner {
             Inner::Valid(open_file_future) => {
-                let (file, mime, chunk_size, maybe_encoding) =
-                    match ready!(Pin::new(open_file_future).poll(cx)) {
-                        Ok(Output::File(file, mime, chunk_size, maybe_encoding)) => {
-                            (file, mime, chunk_size, maybe_encoding)
-                        }
-
-                        Ok(Output::Redirect(location)) => {
-                            let res = Response::builder()
-                                .header(http::header::LOCATION, location)
-                                .status(StatusCode::TEMPORARY_REDIRECT)
-                                .body(empty_body())
-                                .unwrap();
-                            return Poll::Ready(Ok(res));
-                        }
-
-                        Ok(Output::NotFound) => {
-                            let res = Response::builder()
-                                .status(StatusCode::NOT_FOUND)
-                                .body(empty_body())
-                                .unwrap();
-
-                            return Poll::Ready(Ok(res));
+                let response = match ready!(Pin::new(open_file_future).poll(cx)) {
+                    Ok(Output::File(file_request)) => {
+                        let mut builder = Response::builder()
+                            .header(header::CONTENT_TYPE, file_request.mime_header_value)
+                            .header(header::ACCEPT_RANGES, "bytes")
+                            .header(header::CONTENT_LENGTH, file_request.total_size.to_string());
+                        if let Some(encoding) = file_request.maybe_encoding {
+                            builder = builder
+                                .header(header::CONTENT_ENCODING, encoding.into_header_value());
                         }
-
-                        Err(err) => {
-                            return Poll::Ready(
-                                super::response_from_io_error(err)
-                                    .map(|res| res.map(ResponseBody::new)),
+                        if let Some(range) = file_request.maybe_range {
+                            match range {
+                                ParsedRangeHeader::Range(ranges) => {
+                                    let range = &ranges[0];
```

Comment:
Is there a way to write this such it'll never panic? I'm always scared of stuff like this if we refactor the code someday in a way that `ranges` can be empty here.